mathics/builtin/tensors.py

```diff
@@ -329,7 +362,62 @@ def rec(item, rest_lists, current):
                     elements.append(rec(element, rest_lists, current))
                 return Expression(head, *elements)
 
-        return rec(lists[0], lists[1:], [])
+        def rec_sparse(item, rest_lists, current):
```

Comment:
Maybe this code deserves to be moved to a  `mathics.eval` module.

Reply:
Could you tell me more about how the codes are organized in Mathics3? So that I can determine which code should be moved to ``mathics.eval`` or other locations in future updates.

Reply:
This aspect is described in https://mathics-development-guide.readthedocs.io/en/latest/extending/code-overview/python-modules.html and `mathics.eval` specifically in https://mathics-development-guide.readthedocs.io/en/latest/extending/code-overview/python-modules.html

Reply:
Thanks for the references. But I still have some specific questions left to figure out
1. Should ``mathics.eval`` be as general as possible, or is it just a place for ``eval()`` of functions in ``mathics.builtin``?
2. I don't feel any essential difference between ``rec()`` and ``rec_sparse()``. So if ``rec_sparse()`` should be moved to ``mathics.eval.tensor``, I'd prefer to treat ``rec()`` (in ``Outer`` and in ``Inner``) the same way.
3. If ``rec()`` is moved to ``mathics.eval.tensor``, it will require as many as 6 parameters ``(item, rest_lists, current, f, head, evaluation)``, which seems a bit weird to me.

Reply:
The question as is leads me to believe that I  might not have have conveyed the motivation. So let me start with that and wind up with some guidelines which I hope you can then decide on how to organize.

Currently Mathics3 evaluates expressions by walking a tree (which is specifically a kind of M-expression)

If Mathics3 is to be sped up, we need to turn the tree evaluation into a sequence of instructions. The functions that start `eval_`  you can think of as the instructions. They are top-level functions like Add, Subtract, Factor, etc. but with parameter checking done and objects like Integers and Strings turned into possibly lower, level objects. Basically there would be at least one for every Wolfram Alpha builtin function. 

Of course, to support these high-level built in functions, we need other functions. Those will _not_ start with `eval_`.  

With respect to `rec()` and `rec_sparse()` specifically, if you do not feel any essential difference between them, then, sure, they both can be moved somewhere under the the `mathics.eval` module. Please, can we we come up with a more descriptive name than `rec()` and `rec_sparse()` and add docstrings to these functions to as well? 

So finally, I have to describe or apologize for the crappiness of this code, since we are trying to encourage people not to follow the bad habits of what came before. 

Originally, this was one huge monolith and we have not come close to digging ourselves out of the mess that is there. In the hierarchical OO monolith, there were a number of nested private functions such as `rec()` (or is it "wreck") that had to duplicated in that organization and because the were or are, well, private functions which probably should not have been private but instead put in a common module where the function can be used both places. 

Finally as for the number of parameters.  It probably means that there is a structure, object, concept or organization that we are missing that exists that encapsulates those parts into a more comprehensive whole. 

However since I don't understand what this `rec` thing does, clearly I can't offer a suggestion.  If you have an idea here and want to go with that, please suggest something or make that change. 

Otherwise for the first cut, I would not worry about this initially. We can come back to that. What I have found is that there is so much of a mess, we just can't fix everything all at once. And whatever the ultimate better solution is, having splitting off the common `rec` thing into a separate module will only help things.

Reply:
@Li-Xiang-Ideal now that the CI fixes are in, please rebase this PR so we don't have the duplication.

Reply:
I looked at the situation with the `rec()` and `rec_sparse()` in more detail. Some of the things I said were a bit misleading, so let me clarify the situation here. 

First, yes indeed, this code goes back to the earliest times and that means the coding style is not good; names are misleading and there are neither docstrings nor documentation describing the code.

I suppose "rec" relates to the recursive implementation. The only way one could make this more vague would to use `f()` or `fn()`.  But thankfully, at least the _wrong_ term has not be used which was sometimes done.

In good programming practice, function names reflect what a function does more than how it is implemented. And  here the name is vague too.

As a result, you find a number of internal functions called `rec()`  in different Builtin classes that are really not related to one another. But the two functions we are interested in here which are similar are in `rec` and `rec_sparse` from Builtin `Outer`.  

A better name for these two functions would be `create_outer_piece()` and `create_outer_piece_sparse()`. If it is useful to mention these work recursively, that would be done in the docstring or _description_ of the function. For example: 

```
"""Recursively computes the outer product of list `item`  with `rest_lists`; `current` contains the results computed so far.
"""
```

It might also be nice to add type annotations. (This code was written in Python 2.7 when type annotations did not exist. But there comments, docstrings and good programming practice _did_ still exist in Python 2.7.) 

Here is a suggestion for reorganization...

 Create a  _function_ called `eval_Outer()`  in `mathics.eval.tensors` ,  and move the current  method in `Outer` called `eval()` to that renaming and improving the `rec()` and `rec_sparse()` methods.  They can still be an internal function as it is here. 

 `eval()` then just calls `eval_Outer` .

Later, if we want reorganized  things further we are in a better position to do that.

I suspect that you like most people were expecting you'd just make this little change and to get the code to understand sparse arrays and be done with it.  I know if I were contributing here, that's what I expect.

The reality is there is a big mess that we picked up that we are still trying to dig ourselves out of.  In the past most of our effort has been just clean up work of the kind mentioned here. 

It has been a big mistake to just keep coding in the style  and organization that came before. That's what led to this being a _big_ mess rather than just a mess.

To the extent you can help us dig out of this, this is greatly apprecated.

Reply:
I have to say, my main concern was about having a more or less complicated implementation of the `eval_` methods inside the `mathics.builtin.tensor` module, not about the private, nested functions. Of course, it is better if these functions are properly named and documented, but at this stage, it would be enough for me to move the implementation to `mathics.eval.tensor`, something like


``` # In mathics.eval.tensor.py

def eval_outer(f, list, evaluation):
    """Evaluates recursively the outer product"""
    ...
```

and 

In mathics.builtin.tensor
```

    def eval(self, f, lists, evaluation):
        "Outerf[f_, lists__]"
        return eval_outer(f, lists, evaluation)
```

Reply:
Yes, I guess so.  I suppose this would be a good way to describe it in the developer documentation.

And like C/C++ header-source separation it is not strictly mandatory to make code work, just accepted practice.

As @mmatera says, this is good.  We can merge this as is, or do you want to do the other stuff in addition. That can be done in another PR if you prefer.